server/auth/auth.go

```diff
@@ -0,0 +1,740 @@
+	if err != nil {
+		return nil, err
+	}
+
+	var user *User
+	if userJSON == "" {
+		user = nil
+	} else {
+		user, err = NewUserFromJSON(userJSON)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// If another (write) operation has already updated the cache,
+	// we abandon the data we've just retrieved and use the cached version.
+	_latestUser, _ := m.users.LoadOrStore(name, user)
+	latestUser := _latestUser.(*User)
+
+	if latestUser == nil {
+		return nil, ErrUserNotFound(name)
+	}
+	return latestUser, nil
+}
+
+// GetUsers returns all available users.
+func (m *Manager) GetUsers() (map[*User]struct{}, error) {
+	// NOTE: GetUsers doesn't use the cache and read data from kv directly to avoid cache inconsistency.
+	usersPath := strings.Join([]string{userPrefix, ""}, "/")
+	_, _users, err := m.kv.LoadRange(usersPath, clientv3.GetPrefixRangeEnd(usersPath), 0)
+	if err != nil {
+		return nil, err
+	}
+
+	users := make(map[*User]struct{})
+	for _, userJSON := range _users {
+		user, err := NewUserFromJSON(userJSON)
+		if err != nil {
+			return nil, err
+		}
+
+		users[user] = struct{}{}
+	}
+
+	return users, nil
+}
+
+// GetUserKeys returns names of all available users.
+func (m *Manager) GetUserKeys() (map[string]struct{}, error) {
+	// NOTE: GetRoleKeys doesn't use the cache and read data from kv directly to avoid cache inconsistency.
+	usersPath := strings.Join([]string{userPrefix, ""}, "/")
+	_keys, _, err := m.kv.LoadRange(usersPath, clientv3.GetPrefixRangeEnd(usersPath), 0)
+	if err != nil {
+		return nil, err
+	}
+
+	keys := make(map[string]struct{})
+	for _, key := range _keys {
+		keys[key] = struct{}{}
+	}
+
+	return keys, nil
+}
+
+// CreateUser creates a new user.
+func (m *Manager) CreateUser(name string, password string) error {
+	_, err := m.GetUser(name)
+	if err == nil {
+		return ErrUserExists(name)
+	}
+	if err.Error() != ErrUserNotFound(name).Error() {
+		return err
+	}
+
+	user, err := NewUser(name, GenerateHash(password))
+	if err != nil {
+		return err
+	}
+
+	// Try to add user to kv first.
+	userJSON, err := json.Marshal(user)
+	if err != nil {
+		return err
+	}
+	userPath := path.Join(userPrefix, name)
+
+	_, err = m.kv.NewTxn().
+		If(kv.Eq(kv.Create(userPath), 0)).
+		Then(kv.OpSave(name, string(userJSON))).
+		Commit()
+	if err != nil {
+		return err
+	}
+
+	// Add user to memory cache.
+	m.users.Store(name, user)
+
+	return nil
+}
+
+// DeleteUser deletes a user.
+func (m *Manager) DeleteUser(name string) error {
+	user, err := m.GetUser(name)
+	if err != nil {
+		return err
+	}
+
+	// Try to delete user from kv first.
+	userJSON, err := json.Marshal(user)
+	if err != nil {
+		return err
+	}
+	userPath := path.Join(userPrefix, name)
+
+	_, err = m.kv.NewTxn().
+		If(kv.Eq(kv.Value(userPath), string(userJSON))).
+		Then(kv.OpRemove(userPath)).
+		Commit()
+	if err != nil {
+		return err
+	}
+
+	// Delete user from memory cache.
+	m.users.Store(name, (*User)(nil))
```

Comment:
How does the memory cache interact with the PD leader changing? Will it get refreshed when the leader changes?